packages/monitoring/src/serviceMonitor/serviceURLManager.ts

```diff
@@ -0,0 +1,127 @@
+   *
+   * @param {ILivenessChecker} service - An instance of ILivenessChecker that provides methods to check the service's liveness.
+   *
+   * @returns {Promise<{alive: boolean, error?: Error}>} - A promise that resolves with the liveness status of the service.
+   */
+  private async pingService(service: ILivenessChecker) {
+    try {
+      const statusPromise = service.isAlive();
+
+      const timeoutPromise = new Promise((resolve, reject) => {
+        setTimeout(() => {
+          reject(new Error("Timeout"));
+        }, this.timeout * 1000);
+      });
+
+      const result = await Promise.race([statusPromise, timeoutPromise]);
+      if (result instanceof Error && result.message === "Timeout") {
+        throw result;
+      }
+      if ("disconnect" in service) {
+        await (service as IDisconnectHandler).disconnect();
+      }
+      return result;
+    } catch (e) {
+      return this.handleErrorsOnSilentMode((e as Error).message);
+    }
+  }
+  /**
+   * Handles errors based on the silent mode setting.
+   *
+   * If silent mode is enabled, logs the error message and returns null.
+   * If silent mode is disabled, throws an error with the provided message.
+   *
+   * @private
+   * @param {string} errorMsg - The error message to handle.
+   * @returns {null} - Returns null if silent mode is enabled, otherwise throws an error.
+   * @throws {Error} - Throws an error with the provided message if silent mode is disabled.
+   */
+  private handleErrorsOnSilentMode(errorMsg: string) {
+    if (this.silent) {
+      monitorEvents.log(errorMsg, "red");
+      return null;
+    }
+    throw new Error(errorMsg);
+  }
+
+  /**
+   * Attempts to find a reachable service URL from a list of provided URLs.
+   *
+   * This method iterates through the list of URLs, repeatedly pinging the service to
+   * check if it is alive. If a reachable URL is found, it is returned. If all URLs
+   * are exhausted without finding a reachable service, an error is thrown.
+   *
+   * @param {string[]} urls - An array of service URLs to check for reachability.
+   * @param {ILivenessChecker} service - An instance of ILivenessChecker that provides methods
+   * to check the service's liveness and manage service URLs.
+   *
+   * @returns {Promise<string>} - A promise that resolves with the reachable service URL.
+   *
+   * @throws {Error} - Throws an error if no reachable service URL is found after checking all URLs.
+   *
+   */
+  async getAvailableStack(urls: string[], service: ILivenessChecker): Promise<ServiceUrl<N>> {
+    let error: Error | string = "";
+    for (let i = 0; i < urls.length; i++) {
+      await service.update({ url: urls[i] });
+      monitorEvents.log(`${service.name}: pinging ${service.url}`, "gray");
+      for (let retry = 0; retry < this.retries; retry++) {
+        const status = await this.pingService(service);
```

Comment:
let's not wait for each url to respond, we can spawn all the requests simultaneously. and the wait for the promises to resolve (resolve with the first success)